sol/main.c

```diff
@@ -0,0 +1,551 @@
+		insert_new_priority_req(req_queue, req, priority);
+	} else {
+		/* Append request to end of the appropriate priority. */
+		list_add(&req->list, &req_queue->priorities[priority]->list);
+		req_queue->priorities[priority] = req;
+	}
+
+	req_queue->len++;
+}
+
+static void
+enqueue_reqs(struct sol_config *sol_conf)
+{
+	struct priority_req *req_nodes[sol_conf->enq_burst_size];
+	int num_reqs = mb_dequeue_burst(&sol_conf->mb,
+		(void **)req_nodes, sol_conf->enq_burst_size);
+	int i;
+	for (i = 0; i < num_reqs; i++)
+		enqueue_req(sol_conf, req_nodes[i]);
+}
+
+static inline void
+credits_update(struct req_queue *req_queue)
+{
+	uint64_t curr_cycles = rte_rdtsc();
+	uint64_t avail_cycles = curr_cycles - req_queue->time_cpu_cycles;
+#if __WORDSIZE == 64
+	ldiv_t avail_bytes;
+#else
+	lldiv_t avail_bytes;
+#endif
+
+	/*
+	 * Not enough cycles have passed to update the number of credits.
+	 *
+	 * XXX We could make this more general -- i.e. only proceed with
+	 * the division below if enough cycles have passed to add 64 bytes or
+	 * a mimimum packet size, etc.
+	 */
+	if (avail_cycles <= req_queue->cycles_per_byte_floor)
+		return;
+
+#if __WORDSIZE == 64
+	avail_bytes = ldiv(avail_cycles * req_queue->cycles_per_byte_a,
+		req_queue->cycles_per_byte_b);
+#else
+	avail_bytes = lldiv(avail_cycles * req_queue->cycles_per_byte_a,
+		req_queue->cycles_per_byte_b);
+#endif
+
+	req_queue->tb_credit_bytes += avail_bytes.quot;
+	if (req_queue->tb_credit_bytes > req_queue->tb_max_credit_bytes)
+		req_queue->tb_credit_bytes = req_queue->tb_max_credit_bytes;
+
+	/*
+	 * If there are spare cycles (that were not converted to credits
+	 * because of rounding), keep them for the next iteration.
+	 */
+	req_queue->time_cpu_cycles = curr_cycles - avail_bytes.rem;
+}
+
+static inline int
+credits_check(struct req_queue *req_queue, struct rte_mbuf *pkt)
+{
+	/* Need to include Ethernet frame overhead (preamble, gap, etc.) */
+	uint32_t pkt_len = pkt->pkt_len + RTE_SCHED_FRAME_OVERHEAD_DEFAULT;
+	if (pkt_len > req_queue->tb_credit_bytes)
+		return false;
+	req_queue->tb_credit_bytes -= pkt_len;
+	return true;
+}
+
+static void
+dequeue_reqs(struct sol_config *sol_conf, uint8_t tx_port)
+{
+	struct req_queue *req_queue = &sol_conf->req_queue;
+	struct priority_req *entry, *next;
+
+	struct rte_mbuf *pkts_out[sol_conf->deq_burst_size];
+	uint32_t nb_pkts_out = 0;
+	uint16_t total_sent = 0;
+
+	/* Get an up-to-date view of our credits. */
+	credits_update(req_queue);
+
+	list_for_each_entry_safe(entry, next, &req_queue->head, list) {
+		struct rte_mbuf *pkt = entry->pkt;
+
+		if (!credits_check(req_queue, pkt)) {
+			/*
+			 * XXX When under an attack, we may not want to
+			 * log this because it could become expensive.
+			 */
+			RTE_LOG(NOTICE, GATEKEEPER,
+				"sol: out of request bandwidth\n");
+			goto out;
+		}
+
+		if (req_queue->len == 1 || (entry->priority != next->priority))
+			req_queue->priorities[entry->priority] = NULL;
```

Comment:
If `entry` is the lowest priority packet, then `next` is the `list_entry()` of the head of the list, which is not a valid pointer to a `struct priority_req`.

Putting the test `req_queue->len == 1` first is enough to protect against `next` being used when it's not a valid pointer, but I could see how that could be perceived as sloppy. I can't find any examples of using `list_for_each_entry_safe()` where the temporary storage parameter (`next` in this case) is dereferenced, perhaps for this reason.

Should I instead keep a variable like `prev_priority` and update it in the loop to do an equivalent test without dereferencing `next`?